File: src/core/Cryptos/OwnedCryptoState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using core.Shared;

namespace core.Cryptos
{
    public class OwnedCryptoState : IAggregateState
	{
        public Guid Id { get; private set; }
        public string Token { get; private set; }
        public Guid UserId { get; private set; }

        public decimal Quantity { get; private set; }
        public decimal Cost { get; private set; }
        public decimal AverageCost => Cost / Quantity;
        public List<CryptoTransaction> Transactions { get; } = new();
        internal List<ICryptoTransaction> BuyOrSell { get; } = new();
        internal HashSet<Guid> Deletes { get; } = new();

        public List<PositionInstance> PositionInstances { get; } = new();

        public string Description => $"{Quantity} tokens at avg cost {Math.Round(AverageCost, 2)}";

        public int DaysHeld { get; private set; }
        public int DaysSinceLastTransaction { get; private set; }
        public IEnumerable<AggregateEvent> UndeletedBuysOrSells =>
            BuyOrSell.Where(a => Deletes.Contains(a.Id) == false).Cast<AggregateEvent>();

        public List<CryptoAwarded> Awards { get; } = new List<CryptoAwarded>();
        public List<CryptoYielded> Yields { get; } = new List<CryptoYielded>();

        internal void ApplyInternal(CryptoObtained o)
        {
            Id = o.AggregateId;
            Token = o.Token;
            UserId = o.UserId;
        }

        internal void ApplyInternal(CryptoPurchased purchased)
        {
            BuyOrSell.Add(purchased);

            StateUpdateLoop();
        }

        internal void ApplyInternal(CryptoDeleted deleted)
        {
            foreach(var t in BuyOrSell)
            {
                Deletes.Add(t.Id);
            }

            StateUpdateLoop();
        }

        internal void ApplyInternal(CryptoTransactionDeleted deleted)
        {
            Deletes.Add(deleted.TransactionId);

            StateUpdateLoop();
        }

        internal void ApplyInternal(CryptoSold sold)
        {
            BuyOrSell.Add(sold);

            StateUpdateLoop();
        }

        internal void ApplyInternal(CryptoAwarded awarded)
        {
            Awards.Add(awarded);

            StateUpdateLoop();
        }

        internal void ApplyInternal(CryptoYielded yielded)
        {
            Yields.Add(yielded);

            StateUpdateLoop();
        }

        private void StateUpdateLoop()
        {
            decimal quantity = 0;
            decimal cost = 0;
            var txs = new List<CryptoTransaction>();
            DateTimeOffset? oldestOpen = null;
            var positionInstances = new List<PositionInstance>();
            DateTimeOffset lastTransaction = DateTimeOffset.UtcNow;

            bool PurchaseProcessing(ICryptoTransaction st)
            {
                lastTransaction = st.When;

                if (quantity == 0)
                {
                    oldestOpen = st.When;
                    positionInstances.Add(new PositionInstance(Token));
                }

                quantity += st.Quantity;
                cost += st.DollarAmount;

                txs.Add(
                    CryptoTransaction.DebitTx(
                        Id,
                        st.Id,
                        Token,
                        $"Purchased {st.Quantity} for ${st.DollarAmount}",
                        st.DollarAmount / st.Quantity,
                        st.DollarAmount,
                        st.When
                    )
                );

                positionInstances[positionInstances.Count - 1].Buy(st.Quantity, st.DollarAmount, st.When);

                return true;
            }

            bool SellProcessing(ICryptoTransaction st)
            {
                // TODO: this should never happen but in prod I see sell get
                // triggered before purchase... something is amiss
                if (positionInstances.Count > 0)
                    positionInstances[positionInstances.Count - 1].Sell(st.Quantity, st.DollarAmount, st.When);

                lastTransaction = st.When;

                txs.Add(
                    CryptoTransaction.CreditTx(
                        Id,
                        st.Id,
                        Token,
                        $"Sold {st.Quantity} for ${st.DollarAmount}",
                        st.DollarAmount / st.Quantity,
                        st.DollarAmount,
                        st.When
                    )
                );
                
                quantity -= st.Quantity;
                cost -= st.DollarAmount;

                return true;
            }

            foreach(var st in BuyOrSell.OrderBy(e => e.When).ThenBy(i => BuyOrSell.IndexOf(i)))
            {
                if (Deletes.Contains(st.Id))
                {
                    continue;
                }

                if (st is CryptoPurchased sp)
                {
                    PurchaseProcessing(sp);
                }
                else if (st is CryptoSold ss)
                {
                    SellProcessing(ss);
                }
                
                if (quantity == 0)
                {
                    cost = 0;
                    oldestOpen = null;
                }
            }

            foreach(var a in Awards)
            {
                if (Deletes.Contains(a.Id))
                {
                    continue;
                }

                if (quantity == 0)
                {
                    oldestOpen = a.When;
                }

                quantity += a.Quantity;
            }

            foreach(var y in Yields)
            {
                if (Deletes.Contains(y.Id))
                {
                    continue;
                }

                quantity += y.Quantity;
            }

            Quantity = quantity;
            Cost = cost;
            Transactions.Clear();
            Transactions.AddRange(txs);
            PositionInstances.Clear();
            PositionInstances.AddRange(positionInstances);

            DaysHeld = oldestOpen.HasValue ? 
                (int)Math.Floor(DateTimeOffset.UtcNow.Subtract(oldestOpen.Value).TotalDays) : 0;

            DaysSinceLastTransaction = (int)DateTimeOffset.UtcNow.Subtract(lastTransaction).TotalDays;
        }

        public void Apply(AggregateEvent e)
        {
            ApplyInternal(e);
        }

        protected void ApplyInternal(dynamic obj)
        {
            ApplyInternal(obj);
        }
    }
}

File: src/core2.fs/Cryptos/OwnedCryptoState.fs
namespace core.Cryptos

open System
open System.Collections.Generic
open System.Linq
open core.Shared

type OwnedCryptoState() =
    let mutable _id = Guid.Empty
    let mutable _token: string = null
    let mutable _userId = Guid.Empty
    let mutable _quantity = 0m
    let mutable _cost = 0m
    let _transactions = List<CryptoTransaction>()
    let _buyOrSell = List<ICryptoTransaction>()
    let _deletes = HashSet<Guid>()
    let _positionInstances = List<PositionInstance>()
    let _awards = List<CryptoAwarded>()
    let _yields = List<CryptoYielded>()
    let mutable _daysHeld = 0
    let mutable _daysSinceLastTransaction = 0

    member _.Id = _id
    member _.Token = _token
    member _.UserId = _userId
    member _.Quantity = _quantity
    member _.Cost = _cost
    member this.AverageCost = _cost / _quantity
    member _.Transactions = _transactions :> IList<_>
    member internal _.BuyOrSell = _buyOrSell :> IList<_>
    member internal _.Deletes = _deletes :> HashSet<_>
    member _.PositionInstances = _positionInstances :> IList<_>
    member this.Description = sprintf "%M tokens at avg cost %M" _quantity (Math.Round(this.AverageCost, 2))
    member _.DaysHeld = _daysHeld
    member _.DaysSinceLastTransaction = _daysSinceLastTransaction
    member this.UndeletedBuysOrSells =
        _buyOrSell.Where(fun a -> not (_deletes.Contains(a.Id))).Cast<AggregateEvent>()
    member _.Awards = _awards :> IList<_>
    member _.Yields = _yields :> IList<_>

    member internal this.ApplyInternal(o: CryptoObtained) =
        _id <- o.AggregateId
        _token <- o.Token
        _userId <- o.UserId

    member internal this.ApplyInternal(purchased: CryptoPurchased) =
        _buyOrSell.Add(purchased)
        this.StateUpdateLoop()

    member internal this.ApplyInternal(deleted: CryptoDeleted) =
        for t in _buyOrSell do
            _deletes.Add(t.Id) |> ignore
        this.StateUpdateLoop()

    member internal this.ApplyInternal(deleted: CryptoTransactionDeleted) =
        _deletes.Add(deleted.TransactionId) |> ignore
        this.StateUpdateLoop()

    member internal this.ApplyInternal(sold: CryptoSold) =
        _buyOrSell.Add(sold)
        this.StateUpdateLoop()

    member internal this.ApplyInternal(awarded: CryptoAwarded) =
        _awards.Add(awarded)
        this.StateUpdateLoop()

    member internal this.ApplyInternal(yielded: CryptoYielded) =
        _yields.Add(yielded)
        this.StateUpdateLoop()

    member private this.StateUpdateLoop() =
        let mutable quantity = 0m
        let mutable cost = 0m
        let txs = List<CryptoTransaction>()
        let mutable oldestOpen: DateTimeOffset option = None
        let positionInstances = List<PositionInstance>()
        let mutable lastTransaction = DateTimeOffset.UtcNow

        let purchaseProcessing (st: ICryptoTransaction) =
            lastTransaction <- st.When

            if quantity = 0m then
                oldestOpen <- Some st.When
                positionInstances.Add(PositionInstance(_token))

            quantity <- quantity + st.Quantity
            cost <- cost + st.DollarAmount

            txs.Add(
                CryptoTransaction.DebitTx(
                    _id,
                    st.Id,
                    _token,
                    sprintf "Purchased %M for $%M" st.Quantity st.DollarAmount,
                    st.DollarAmount / st.Quantity,
                    st.DollarAmount,
                    st.When
                )
            )

            positionInstances.[positionInstances.Count - 1].Buy(st.Quantity, st.DollarAmount, st.When)
            true

        let sellProcessing (st: ICryptoTransaction) =
            if positionInstances.Count > 0 then
                positionInstances.[positionInstances.Count - 1].Sell(st.Quantity, st.DollarAmount, st.When)

            lastTransaction <- st.When

            txs.Add(
                CryptoTransaction.CreditTx(
                    _id,
                    st.Id,
                    _token,
                    sprintf "Sold %M for $%M" st.Quantity st.DollarAmount,
                    st.DollarAmount / st.Quantity,
                    st.DollarAmount,
                    st.When
                )
            )

            quantity <- quantity - st.Quantity
            cost <- cost - st.DollarAmount
            true

        for st in _buyOrSell.OrderBy(fun e -> e.When).ThenBy(fun i -> _buyOrSell.IndexOf(i)) do
            if not (_deletes.Contains(st.Id)) then
                match st with
                | :? CryptoPurchased as sp -> purchaseProcessing sp |> ignore
                | :? CryptoSold as ss -> sellProcessing ss |> ignore
                | _ -> ()

                if quantity = 0m then
                    cost <- 0m
                    oldestOpen <- None

        for a in _awards do
            if not (_deletes.Contains(a.Id)) then
                if quantity = 0m then
                    oldestOpen <- Some a.When
                quantity <- quantity + a.Quantity

        for y in _yields do
            if not (_deletes.Contains(y.Id)) then
                quantity <- quantity + y.Quantity

        _quantity <- quantity
        _cost <- cost
        _transactions.Clear()
        _transactions.AddRange(txs)
        _positionInstances.Clear()
        _positionInstances.AddRange(positionInstances)

        _daysHeld <-
            match oldestOpen with
            | Some oo -> int (Math.Floor(DateTimeOffset.UtcNow.Subtract(oo).TotalDays))
            | None -> 0

        _daysSinceLastTransaction <- int (DateTimeOffset.UtcNow.Subtract(lastTransaction).TotalDays)

    member this.Apply(e: AggregateEvent) =
        this.ApplyInternal(e :> obj)

    member private this.ApplyInternal(obj: obj) =
        match obj with
        | :? CryptoObtained as e -> this.ApplyInternal(e)
        | :? CryptoPurchased as e -> this.ApplyInternal(e)
        | :? CryptoDeleted as e -> this.ApplyInternal(e)
        | :? CryptoTransactionDeleted as e -> this.ApplyInternal(e)
        | :? CryptoSold as e -> this.ApplyInternal(e)
        | :? CryptoAwarded as e -> this.ApplyInternal(e)
        | :? CryptoYielded as e -> this.ApplyInternal(e)
        | _ -> ()

    interface IAggregateState with
        member this.Id = this.Id
        member this.Apply(e) = this.Apply(e)
